File: Native/Android/LunarConsole/lunarConsole/src/main/java/spacemadness/com/lunarconsole/dependency/DependencyProvider.java
package spacemadness.com.lunarconsole.dependency;

import java.util.HashMap;
import java.util.Map;

import static spacemadness.com.lunarconsole.utils.ObjectUtils.checkNotNull;

public class DependencyProvider {
	private static Map<Class<? extends Dependency>, Dependency> dependencyLookup = new HashMap<>();

	public static <T extends Dependency> void register(Class<T> cls, T dependency) {
		dependencyLookup.put(checkNotNull(cls, "cls"), checkNotNull(dependency, "dependency"));
	}

	static void clear() {
		dependencyLookup.clear();
	}

	public static <T extends Dependency> T of(Class<T> cls) {
		Dependency dependency = dependencyLookup.get(cls);
		if (dependency == null) {
			throw new IllegalStateException("Dependency not registered: " + cls);
		}
		//noinspection unchecked
		return (T) dependency;
	}
}

File: Native/Android/LunarConsoleNew/lunarCore/src/main/java/spacemadness/com/lunarconsole/di/DependencyProvider.kt
package spacemadness.com.lunarconsole.di

import androidx.annotation.VisibleForTesting

/**
 * Service Locator design pattern implementation. Provides a lazy dependency injection resolution
 * based on the target class.
 */
object DependencyProvider {
    /**
     * Provider lookup by target class.
     */
    private val lookup = mutableMapOf<Class<*>, Provider<*>>()

    /**
     * Registers new [Provider] for a specified class.
     */
    inline fun <reified T : Any> register(provider: Provider<T>) {
        register(T::class.java, provider)
    }

    /**
     * Registers new target object for a specified class.
     */
    inline fun <reified T : Any> register(target: T) {
        register(T::class.java, object : Provider<T> {
            override fun get(): T = target
        })
    }

    /**
     * Registers new [Provider] for a specified class.
     */
    fun <T : Any> register(type: Class<T>, provider: Provider<T>) {
        lookup[type] = provider
    }

    /**
     * Returns provide associated with this target type.
     */
    inline fun <reified T : Any> of() = of(T::class.java)

    /**
     * Returns provide associated with this target type.
     */
    fun <T : Any> of(type: Class<T>): T {
        val provider = lookup[type]
            ?: throw IllegalArgumentException("Provider is not registered: $type")

        @Suppress("UNCHECKED_CAST")
        return provider.get() as T
    }

    /**
     * Clears all dependency providers.
     */
    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    fun clear() {
        lookup.clear()
    }
}